jerry-core/parser/regexp/re-parser.c

```diff
@@ -35,559 +35,760 @@
 {
-  const lit_utf8_byte_t *curr_p = parser_ctx_p->input_curr_p;
+  return (is_capturing) ? RE_OP_CAPTURING_GROUP_START : RE_OP_NON_CAPTURING_GROUP_START;
+} /* re_get_group_start_opcode*/
 
-  if (JERRY_UNLIKELY (curr_p + lookup > parser_ctx_p->input_end_p))
+/**
+ * Get the end opcode for the current group.
+ *
+ * @return RegExp opcode
+ */
+static re_opcode_t
+re_get_group_end_opcode (re_compiler_ctx_t *re_ctx_p, /**< RegExp compiler context */
+                         bool is_capturing) /**< is capturing group */
+{
+  if (is_capturing)
   {
-    return false;
+    if (re_ctx_p->token.greedy)
+    {
+      return RE_OP_GREEDY_CAPTURING_GROUP_END;
+    }
+
+    return RE_OP_LAZY_CAPTURING_GROUP_END;
   }
 
-  for (uint32_t i = 0; i < lookup; i++)
+  if (re_ctx_p->token.greedy)
   {
-    if (!lit_char_is_hex_digit (*curr_p++))
-    {
-      return false;
-    }
+    return RE_OP_GREEDY_NON_CAPTURING_GROUP_END;
   }
 
-  return true;
-} /* re_hex_lookup */
+  return RE_OP_LAZY_NON_CAPTURING_GROUP_END;
+} /* re_get_group_end_opcode */
 
 /**
- * Consume non greedy (question mark) character if present.
- *
- * @return true - if non-greedy character found
- *         false - otherwise
+ * Enclose the given bytecode to a group.
  */
-static inline bool JERRY_ATTR_ALWAYS_INLINE
-re_parse_non_greedy_char (re_parser_ctx_t *parser_ctx_p) /**< RegExp parser context */
+static void
+re_insert_into_group (re_compiler_ctx_t *re_ctx_p, /**< RegExp compiler context */
+                      uint32_t group_start_offset, /**< offset of group start */
+                      uint32_t idx, /**< index of group */
+                      uint32_t capture_start, /**< index of first nested capture */
+                      bool is_capturing) /**< is capturing group */
 {
-  if (parser_ctx_p->input_curr_p < parser_ctx_p->input_end_p
-      && *parser_ctx_p->input_curr_p == LIT_CHAR_QUESTION)
+  uint32_t qmin = re_ctx_p->token.qmin;
+  uint32_t qmax = re_ctx_p->token.qmax;
+
+  if (JERRY_UNLIKELY (!is_capturing && re_bytecode_size (re_ctx_p) == group_start_offset))
+  {
+    return;
+  }
+
+  if (qmin == 0)
   {
-    parser_ctx_p->input_curr_p++;
-    return true;
+    re_insert_value (re_ctx_p,
+                     group_start_offset,
+                     re_bytecode_size (re_ctx_p) - group_start_offset);
   }
 
-  return false;
-} /* re_parse_non_greedy_char */
+  re_insert_value (re_ctx_p, group_start_offset, qmin);
+  re_insert_value (re_ctx_p, group_start_offset, re_ctx_p->captures_count - capture_start);
+
+  if (!is_capturing)
+  {
+    re_insert_value (re_ctx_p, group_start_offset, capture_start);
+  }
+  else
+  {
+    JERRY_ASSERT (idx == capture_start);
+  }
+
+  re_insert_value (re_ctx_p, group_start_offset, idx);
+  re_insert_opcode (re_ctx_p, group_start_offset, re_get_group_start_opcode (is_capturing));
+
+  re_append_opcode (re_ctx_p, re_get_group_end_opcode (re_ctx_p, is_capturing));
```

Comment:
How about reserving space before these 4 appends?

Reply:
I don't think it would improve much, this isn't a performance critical code path. Precalculating the required size would also increase complexity due to how values are encoded.